### ISSUE TITLE
bytestreamToUMP handles reserved status codes inconsistently

### ISSUE THREAD

**Issue by paulhuggett** [midi2-dev/AM_MIDI2.0Lib#17]
I have noticed that handling of the MIDI reserved status codes is somewhat inconsistent. This table shows some test cases showing the current behaviour:

| Input Bytestream | bytestreamToUMP MIDI1 Output | bytestreamToUMP MIDI2 Output |
| ----- | ------ | ----- |
| 0xF4, 0x01, 0x02, 0x03, 0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000 |
| 0xF5, 0x01, 0x02, 0x03, 0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000 |
| 0xF9, 0x01, 0x02, 0x03, 0x91, 0x3C, 0x7F | 0x10F90000, 0x20913C7F | 0x10F90000, 0x40913C00, 0xFFFF0000 |
| 0xFD, 0x01, 0x02, 0x03, 0x91, 0x3C, 0x7F | 0x10FD0000, 0x20913C7F | 0x10FD0000, 0x40913C00, 0xFFFF0000

0xF4 and 0xF5 status codes and following data bytes are ignored and we recover at the start of the next status code. Codes 0xF9 and 0xFD have been converted to undefined UMP messages. I'm not sure that this is the intended behaviour.

I think that the problem is that `bytestreamToUMP::bytestreamParse()` treats 0xF9 and 0xFD as system real-time messages. A simple fix makes the handling of 0xF9/0xFD match that of 0xF4/0xF5. 

Change bytestreamToUMP.h from:
~~~cpp
void bytestreamParse(uint8_t midi1Byte){
  if (midi1Byte == TUNEREQUEST || midi1Byte >=  TIMINGCLOCK) {
    d0 = midi1Byte;
    bsToUMP(midi1Byte,0,0);
    return;
  }
…
~~~
to something like:
~~~cpp
private:
  constexpr bool isSystemRealTimeMessage (std::uint8_t const midi1Byte) {
    switch (midi1Byte) {
    case SEQSTART:
    case SEQCONT:
    case SEQSTOP:
    case ACTIVESENSE:
    case SYSTEMRESET:
      return true;
    default:
      return false;
    }
  }
public:
  void bytestreamParse(std::uint8_t midi1Byte) {
    if (midi1Byte == TIMINGCLOCK || isSystemRealTimeMessage (midi1Byte)) {
      d0 = midi1Byte;
      return this->bsToUMP(midi1Byte, 0, 0);
    }
…
~~~

All the existing tests pass. I'm happy to submit a PR for this change (I hesitate because I don't want to swamp you with more of them)!

A footnote to decode some of the hex in the table (possibly only for my benefit):

- 0xF4, 0xF5, 0xF9, and 0xFD are all reserved and undefined status codes
- 0x91, 0x3C, 0x7F is note-on (channel 1, note 60, velocity 127) expressed in a byte-stream
- 0x20913C7F is a MIDI1 UMP form for the same note-on event
- 0x40913C00, 0xFFFF0000 is a MIDI2 UMP form again for the same event

**Comment by starfishmod** [comment]
@paulhuggett thanks for the detailed analysis. This also relates to #16 

I have been discussing this with some collegues in the MA. My general view is that 0xF4, 0xF5, 0xF9, and 0xFD should not be translated at all and instead the data should be dropped. This is because there is no defined length for these bytes - and technically these messages don't exist in UMP.

I will be looking at this fix - unless you want to beat me to it :laughing:

**Comment by paulhuggett** [comment]
@starfishmod thanks for the response. 

TLDR: I think I might have already implemented the fix you suggest above! 😀

I agree that the right thing for the library to do is drop the reserved status and data bytes. Is there anything else it could reasonably do? AFAICT, the existing code already handles the 0xF4 and 0xF5 codes as you suggest, with only a small tweak required match this behaviour for 0xF9 and 0xFD. 

The table I included above showed inputs `x, 0x01, 0x02, 0x03, 0x91, 0x3C, 0x7F` (where x is each of the four undefined status codes). This represents: the status code (x), three data bytes (0x01, 0x02, 0x03), and a valid note-on message (0x1, 0x3C, 0x7F). The note-on is there to ensure that we recover correctly. The expected output is, of course, just the valid note-on message.

I should probably have also included a table which shows the behaviour with my suggested change in place:

| Input Bytestream | bytestreamToUMP MIDI1 Output | bytestreamToUMP MIDI2 Output |
| ----- | ------ | ----- |
| 0xF4, 0x01, 0x02, 0x03,<br>0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000 |
| 0xF5, 0x01, 0x02, 0x03,<br>0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000 |
| 0xF9, 0x01, 0x02, 0x03,<br>0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000 |
| 0xFD, 0x01, 0x02, 0x03,<br>0x91, 0x3C, 0x7F | 0x20913C7F | 0x40913C00, 0xFFFF0000

The output has dropped the reserved status code and now contains only the note-on message. Just what we're looking for! That's unless I've misunderstood something, of course… 🤔

**Comment by Psychlist1972** [comment]
Agreed on recommended approach to drop the offending status (and data bytes, if provided).

**Comment by starfishmod** [comment]
thanks @paulhuggett. All these undefined/reservered system status codes are now dropped. This ended up being a hefty discussion in the background!

**Comment by paulhuggett** [comment]
Thanks for this. However... commit 122cd2a modifies the umpToBytestream class. This issue describes a very similar problem in the bytestreamToUMP class so I don't think that that is fixed yet!

**Comment by starfishmod** [comment]
Yes sorry there has been a range of commits, and I didn't reference it correctly - however it is no covered in bytestreamToUMP as well :)
